Physical_Engine/Query/IsWithinPileCap.cs

```diff
@@ -0,0 +1,117 @@
+        [Input("padFoundation", "The PadFoundation that defines the extents for the Piles to be located.")]
+        [Input("piles", "The Piles to check if they are located within the extents of the PadFoundation.")]
+        [Input("tolerance", "Distance tolerance to be used in the method. Piles are deemed to be on the edge of the PadFoundation if they are within this distance from the curve.", typeof(Length))]
+        [Output("withinPileCap", "True if all Piles are contained within the boundary and depth of the PileCap. False if one or more Piles are not.")]
+        public static bool IsWithinPileCap(this PadFoundation padFoundation, List<Pile> piles, double tolerance = Tolerance.Distance)
+        {
+            if (padFoundation.IsNull())
+                return false;
+            else if (piles.IsNullOrEmpty())
+                return false;
+            else if (piles.Any(x => x.IsNull()))
+                return false;
+            else if (padFoundation.Construction.IThickness() == 0)
+            {
+                Base.Compute.RecordError("The PadFoundation does not contain a thickness, the method will return false.");
+                return false;
+            }
+
+            // Get the top of the piles
+            List<Point> tops = piles.Select(p => p.Location.IControlPoints().OrderBy(pt => pt.Z).Last()).ToList();
+
+            // Project all points to top surface of the pad
+            Plane padTop = padFoundation.Location.FitPlane();
+            List<Point> pTops = tops.Select(x => x.Project(padTop)).ToList(); //project points only used for IsContaining
+
+
+            // Get the thickness of the PadFoundation to compare against the distance coordinates of the pile tops
+            double padThickness = padFoundation.Construction.IThickness();
+            for (int i = 0; i < tops.Count; i++)
+            {
+                if (tops[i].Distance(pTops[i]) + tolerance > padThickness)
+                {
+                    Base.Compute.RecordError("One or more the Piles tops is not located within the depth of the PileCap.");
+                    return false;
+                }
+            }
+
+            ICurve topOutline = padFoundation.Location.ExternalBoundary;
+
+            // Check if all piles are within the curve of the pad
+            if (topOutline.IIsContaining(pTops, true, tolerance))
+            {
+                    // Check if the profile pushes the pile outside the curve of the pad
+                    for (int i = 0; i < pTops.Count; i++)
+                    {
+                        ConstantFramingProperty property = (ConstantFramingProperty)piles[i].Property;
+                        ICurve profile = Engine.Geometry.Compute.IJoin(property.Profile.Edges.ToList()).OrderBy(c => c.Length()).Last()
+                            .ITranslate(Engine.Geometry.Create.Vector(pTops[i])).IRotate(pTops[i], Vector.ZAxis, property.OrientationAngle);
```

Comment:
As discussed offline with @peterjamesnugent, we should not be doing translations and rotations in here, should be a centralised returning `TransformMatrix`, same as in structures:
https://github.com/BHoM/BHoM_Engine/blob/474c6194f5107f73a0dc0273c06ce8bb47f67744/Structure_Engine/Query/BarSectionTranformation.cs#L43

As discussed, we will rather keep the code as is and sort it out in a separate issue/PR thread, as long as @peterjamesnugent does not change his mind 😉

Reply:
https://github.com/BHoM/BHoM_Engine/issues/3209